Improved Dockerfile and run script

quelsan opened the pull request:
This pull request contains various changes to the Dockerfile and run script. Feel free to cherry pick changes as you deem appropriate.

## TL;DR:
- Allow local modifications of Gophish source code without rewriting import statements
- Include minimization process for client side assets (JavaScript)
- Run container as unprivileged user
- Add option to specify database path

quelsan committed:
Updated Dockerfile and run script

The Dockerfile specifies two separate build containers - one for the
client side JavaScript dependencies, which needs to be minified if
changed, and a build container for Golang.

The Dockerfile has also been rebuilt to use an unprivileged user instead
of root.

The run script adds more options and changes the method the
configuration is overwritten, which may help in situations where the
configuration is not owned by the unprivileged user.

jordan-wright commented:
Hey there!

Thanks for sending this in. I'm excited to get these changes merged. For a while now, I've wanted to improve the Dockerfile so I appreciate you taking this first pass.

> Allow local modifications of Gophish source code without rewriting import statements

Could you expand a bit on this? I'm not seeing where in the changes this has been improved.

jordan-wright commented on docker/run.sh:
```diff
@@ -5,61 +5,69 @@ if [ -n "${ADMIN_LISTEN_URL+set}" ] ; then
         --argjson ADMIN_USE_TLS "${ADMIN_USE_TLS}" \
         '.admin_server.use_tls = $ADMIN_USE_TLS' config.json > config.json.tmp && \
-        mv config.json.tmp config.json
+        cat config.json.tmp > config.json
```
Just curious, why did you take this approach?

quelsan replied:
It is a work around for situations where the configuration is writable by the user but not necessarily owned by it. In this specific case however, if things are run using the shipped Dockerfile, it should not matter.

quelsan commented:
> Hey there!
> 
> Thanks for sending this in. I'm excited to get these changes merged. For a while now, I've wanted to improve the Dockerfile so I appreciate you taking this first pass.
> 
> > Allow local modifications of Gophish source code without rewriting import statements
> 
> Could you expand a bit on this? I'm not seeing where in the changes this has been improved.

"WORKDIR /go/src/github.com/gophish/gophish" is the solution I came up with.
Before go tries to fetch dependencies, it checks if they are already located in the directory.
If the source is not located in this directory, go will download the imported modules from the upstream Github repository.

jordan-wright commented on Dockerfile:
```diff
@@ -1,30 +1,41 @@
-RUN go get -d -v ./...
-RUN go build
 
+RUN npm install gulp gulp-cli -g
```
As a small note, you could move this above the `COPY` line to speed up builds.

jordan-wright commented on Dockerfile:
```diff
@@ -1,30 +1,41 @@
+# Runtime container
+FROM debian:stable-slim
+
+WORKDIR /opt/gophish
```
When running this, I show that gophish is unable to open a SQLite database in /opt/gophish since the folder is owned by root.

It seems a fix for this would be to move this under the `useradd` command, as well as adding the `-m` flag to the `useradd` command in order to have the home directory created.

jordan-wright commented on Dockerfile:
```diff
@@ -1,30 +1,41 @@
+	apt-get install --no-install-recommends -y jq && \
+	apt-get clean && \
+	rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*
+
```
I would move the `USER app` line here, that way files that get copied into the directory are owned by the `app` user.

jordan-wright commented on Dockerfile:
```diff
@@ -1,30 +1,41 @@
+COPY --from=build-golang /go/src/github.com/gophish/gophish/ ./
+COPY --from=build-js /build/static/js/dist/ ./static/js/dist/
+COPY --from=build-js /build/static/css/dist/ ./static/css/dist/
+RUN touch config.json.tmp && chown app. config.json config.json.tmp
```
Once the `USER app` line is moved up, you can remove the `&& chown...` line since it won't be needed anymore.

jordan-wright reviewed (changes requested):
Hey @quelsan,

I was having some issues getting this to work, so I've dropped some comments on how things can probably be tweaked and improved. Give them a shot and let me know what you think!

quelsan commented:
Thanks for the great feedback - I have used a volume mounted DB so I've not experienced the permission issue. Will look into the optimizations and update the branch.

quelsan committed:
Updated Dockerfile with requested changes

quelsan commented:
I've now updated the branch, per your request.

The only difference is that I just moved the copy of the configuration file, as I see no need for the application to be able to modify itself.

quelsan committed:
Fixed permission issue

It seems like Docker COPY does not update permissions if the file has
already been copied, which is the case as we grab the root of the Golang
build.

Therefore, I've added back the chown, which seems to solve the issue.

jordan-wright reviewed (approved):
These look good to me! Thanks again for taking a pass at this @quelsan!

Reedtechno commented:
Looks like this change may have restricted the user down too much:
`time="2019-11-14T02:25:19Z" level=fatal msg="listen tcp 0.0.0.0:80: bind: permission denied"`

icouper-ng commented:
I have run into the issue @Reedtechno identified.

I haven't been able to find instructions on how to elevate the user privileges within the docker container to resolve this. Any suggestions?

jordan-wright commented:
I’ll have a fix for this out soon

> On Nov 20, 2019, at 12:41 PM, icouper-ng <notifications@github.com> wrote:
> 
> ﻿
> I have run into the issue @Reedtechno identified.
> 
> I haven't been able to find instructions on how to elevate the user privileges within the docker container to resolve this. Any suggestions?
> 
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub, or unsubscribe.

icouper-ng commented:
Great to hear. Thanks Jordan!

ACPK commented:
@Reedtechno - I also ran into the same issue after installling the docker issue.

@ jordan-wright - Any updates?
